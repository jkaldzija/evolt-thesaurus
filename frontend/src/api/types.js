export type Word = {
    synonyms: Array<string>,
    value: string
}

export type WordListPage = {
    count: number,
    words: Array<string>
}
