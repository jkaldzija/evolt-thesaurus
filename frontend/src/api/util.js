import Axios, {AxiosInstance} from 'axios'
import {API_URL} from "../config";

export const apiClient: AxiosInstance = Axios.create({
    baseURL: API_URL
});
