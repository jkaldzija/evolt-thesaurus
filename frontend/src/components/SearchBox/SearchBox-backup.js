import React from "react";
import {Button, Card, CardContent, TextField, Typography} from "@material-ui/core";
import './style.css';

type SearchBoxProps = {
    onSubmit: (string) => void
}

export default class SearchBox extends React.Component<SearchBoxProps>{
    state = {
        word: ""
    };

    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit(this.state.word);
    };

    handleChange = e => this.setState({word: e.target.value});

    render() {
        const {word} = this.state;
        return (
            <Card className="search-container">
                <CardContent>
                    <Typography variant="h6" color="default" className="search-title">
                        Search for synonyms
                    </Typography>
                    <form onSubmit={this.onSubmit}>
                        <TextField value={word}
                                   placeholder="Enter word"
                                   fullWidth={true}
                                   onChange={this.handleChange}/>
                        <div className="search-button-container">
                            <Button type="submit" variant="contained" color="primary">Search</Button>
                        </div>
                    </form>
                </CardContent>
            </Card>
        );
    }
}
