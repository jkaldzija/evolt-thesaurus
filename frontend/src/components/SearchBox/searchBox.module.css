

.search-container {
    padding: 24px;
}

.search-title {
    margin-bottom: 12px !important;
}

.button{
    margin-left: 40px;
}

.form{
    margin-bottom:35px;
}

.search-button-container {
    margin-top: 24px;
    display: flex;
    justify-content: center;
}

.search-input{
    border-top: 0px;
    border-left: 0px;
    border-right: 0px;
    border-radius: 0;
}
