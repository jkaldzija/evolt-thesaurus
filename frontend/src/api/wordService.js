import {apiClient} from "./util";
import {AxiosPromise, AxiosResponse} from "axios";
import type {Word, WordListPage} from "./types";

// Define API paths in constants
export const WORD_PATH = "word";
export const WORD_LIST_PATH = "word/list";


export default {

    //List all synonyms for a word
    getSynonymsForWord: (value: string): AxiosPromise<AxiosResponse<Word>> => apiClient.get(WORD_PATH, {
        params: {
            value
        }
    }),

    // Create/Update a word with its synonyms
    createWord: (word: Word): AxiosPromise<AxiosResponse<string>> => apiClient.post(WORD_PATH, word),

    // List the words in the dictionary with pagination
    listWordsPaginated: (offset: number, limit: number): AxiosPromise<AxiosResponse<WordListPage>> => apiClient.get(WORD_LIST_PATH, {
        params: {
            offset,
            limit
        }
    })
}
