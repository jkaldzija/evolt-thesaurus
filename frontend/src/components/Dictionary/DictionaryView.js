import React from 'react';
import {Card, CardBody, CardTitle} from "reactstrap";
import {alphabet} from "../Utility/alphabet";
import wordService from "../../api/wordService";
import WordList from "./WordList";
import {getQueryParam} from "../Utility/routing";
import {getSafe} from "../Utility/state";
import {withRouter} from "react-router";
import {WithLoading} from "../HOC/WithLoading";

const WORDS_PER_PAGE = 10;

const WordListWithLoading = WithLoading(WordList);

// Set prop types
type DictionaryViewProps = {};

class DictionaryView extends React.Component<DictionaryViewProps> {

    constructor(props) {
        super(props);

        this.state = {
            isLoading: true,
            wordList: [],
            count: undefined,
            page: 0
        }
    }

    loadData = async () => {
        this.setState({
            isLoading: true
        } );

        const page = parseInt(getSafe(getQueryParam(this.props.location.search, "page"), 0));

        const {data} = await wordService.listWordsPaginated(page * WORDS_PER_PAGE, WORDS_PER_PAGE);
        this.setState({
            page,
            wordList: data.words,
            wordCount: data.count,
            isLoading: false
        });
    };

    componentDidUpdate(prevProps, prevState){
        if(getQueryParam(prevProps.location.search, "page") != getQueryParam(this.props.location.search, "page")){
            this.loadData();
        }
    }


    componentDidMount(){
        this.loadData();
    }


    render() {
        const {isLoading, wordList, wordCount, page} = this.state;
        return (
            <Card>
                <CardTitle className={"mb-0"}>
                    <h5>Not sure what you are looking for? Take a look at our dictionary:</h5>
                </CardTitle>
                <CardBody className={"px-0"}>
                    <WordListWithLoading
                        isLoading={isLoading}
                        wordList={wordList}
                        pages={wordCount / WORDS_PER_PAGE}
                        page={page}
                    />
                </CardBody>

            </Card>
        );
    };

}

// Set default props
DictionaryView.defaultProps = {};

export default withRouter(DictionaryView);
