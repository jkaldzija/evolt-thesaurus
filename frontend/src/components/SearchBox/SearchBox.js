import React from 'react';
import {Card, Container, Input, Spinner} from "reactstrap";
import style from "./searchBox.module.css";
import classnames from "classnames";
import ThemeButton from "../Form/ThemeButton/ThemeButton";
import {WithLoading} from "../HOC/WithLoading";
import SynonymList from "../SynonymList/SynonymList";
import {withRouter} from "react-router";
import * as qs from "qs";
import wordService from "../../api/wordService";
import {getSafe, getSafeDeep} from "../Utility/state";
import {getQueryParam} from "../Utility/routing";

const SynonymListWithLoading = WithLoading(SynonymList);

class SearchBox extends React.Component {
    constructor(props) {
        super(props);

        let currentSearch = getQueryParam(this.props.location.search, "q");
        if(!currentSearch) currentSearch = "";

        this.state = {
            search: currentSearch,
            currentWord: undefined,
            isLoading: false
        }
    }

    performSearch = async() => {
        if(this.state.search === "") return;
        this.setState({
            isLoading: true
        } );
        let response = {};
        try{
            response = await wordService.getSynonymsForWord(getSafe(this.state.search));
        }catch (e) {

        }
        this.setState({
            currentWord: getSafe(response.data),
            isLoading: false
        });

    };

    componentDidMount(){
        this.performSearch();
    }

    componentDidUpdate(prevProps, prevState, SS){
        const prevSearch = getQueryParam(prevProps.location.search, "q");
        const search = getQueryParam(this.props.location.search, "q");
        console.log(prevSearch + "   " + search);
        if(search != prevSearch){
            this.setState({ search }, () => {
                this.performSearch();
            });

        }

    }


    onSearch = (e) => {
        e.preventDefault();
        let params = qs.parse(this.props.pathname);
        params.q = this.state.search;

        this.props.history.push(`?${qs.stringify(params)}`);

    };


    onInput = e => {
        this.setState({
            [e.target.name]: e.target.value
        })
    };

    reloadData = () => {
        this.performSearch();
    };

    render() {
        return  <Card>
                    <form onSubmit={this.onSearch} className={classnames("d-flex", style.form)}>

                        <Input placeholder={"Please enter a word to find synonims"}
                            className={classnames(style["search-input"])}
                            value={this.state.search}
                            name={"search"}
                            onChange={this.onInput}
                        />
                        <ThemeButton className={style.button}>
                            Search
                        </ThemeButton>

                    </form>
                    <SynonymListWithLoading
                        reloadData={this.reloadData}
                        currentWord={this.state.currentWord}
                        isLoading={this.state.isLoading}
                        synonymList={getSafeDeep(this.state, "currentWord.synonyms", [])}  />


                </Card>
    };

}

export default withRouter(SearchBox);
