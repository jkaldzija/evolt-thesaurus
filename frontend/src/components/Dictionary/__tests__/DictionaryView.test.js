import React from 'react';
import {shallow} from "enzyme";
import DictionaryView from "../DictionaryView";
import {MemoryRouter, Route} from "react-router";
import SearchBox from "../../SearchBox/SearchBox";
import MockAdapter from "axios-mock-adapter";
import {apiClient} from "../../../api/util";
import {WORD_LIST_PATH} from "../../../api/wordService";

describe('Dictionary View', function () {
    it('should render correctly', function () {
        const component = shallow(
            <MemoryRouter>
                <DictionaryView onSubmit={() => {}}/>
            </MemoryRouter>
        ).dive().dive().dive().dive();
        expect(component).toMatchSnapshot();
    });

    it('should load data correctly', async function (){
        const router = shallow(
            <MemoryRouter initialEntries={["/"]}>
                <Route  component={DictionaryView} />
            </MemoryRouter>
        );
        const component = router.dive().dive().dive().dive().dive();

        const response = {
            words: [],
            count: 10
        };
        component.setProps({
            location:{
                search:'?page=1'

            }
        });
        const mock = new MockAdapter(apiClient);
        mock.onGet(WORD_LIST_PATH).reply(200, response);

        await component.instance().loadData();
        expect(component.state("wordCount")).toEqual(response.count);
    });

});
