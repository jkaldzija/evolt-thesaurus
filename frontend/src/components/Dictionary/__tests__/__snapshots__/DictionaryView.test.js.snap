// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionary View should render correctly 1`] = `
<Card
  tag="div"
>
  <CardTitle
    className="mb-0"
    tag="div"
  >
    <h5>
      Not sure what you are looking for? Take a look at our dictionary:
    </h5>
  </CardTitle>
  <CardBody
    className="px-0"
    tag="div"
  >
    <_temp
      isLoading={true}
      page={0}
      pages={NaN}
      wordList={Array []}
    />
  </CardBody>
</Card>
`;
